segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/BookKeeperStorage.java

```diff
@@ -0,0 +1,238 @@
+        ensureInitializedAndNotClosed();
+
+        SegmentProperties segmentProperties = manager.getOrRetrieveStorageLedgerDetails(streamSegmentName, true);
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return segmentProperties;
+    }
+
+    @Override
+    public boolean exists(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "exists", streamSegmentName);
+        ensureInitializedAndNotClosed();
+
+        boolean bool = manager.exists(streamSegmentName);
+        LoggerHelpers.traceLeave(log, "exists", traceId, streamSegmentName);
+        return bool;
+    }
+
+    @Override
+    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        ensureInitializedAndNotClosed();
+
+        manager.fence(streamSegmentName);
+        SegmentHandle retVal = BookKeeperSegmentHandle.writeHandle(streamSegmentName);
+        LoggerHelpers.traceLeave(log, "openWrite", traceId, streamSegmentName);
+        return retVal;
+    }
+
+    @Override
+    public SegmentProperties create(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+        ensureInitializedAndNotClosed();
+
+        manager.create(streamSegmentName);
+        StreamSegmentInformation segmentProperties = StreamSegmentInformation.builder()
+                                                                             .name(streamSegmentName)
+                                                                             .build();
+        LoggerHelpers.traceLeave(log, "create", traceId, streamSegmentName);
+        return segmentProperties;
+    }
+
+    @Override
+    public void write(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentException {
+        ensureInitializedAndNotClosed();
+
+        Timer timer = new Timer();
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+        Preconditions.checkArgument(handle instanceof BookKeeperSegmentHandle, "handle must be instance of BookKeeper segment handle.");
+        int lengthWritten = manager.write(handle.getSegmentName(), offset, data, length);
+        BookKeeperStorageMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+        BookKeeperStorageMetrics.WRITE_BYTES.add(lengthWritten);
+    }
+
+    @Override
+    public void seal(SegmentHandle handle) throws StreamSegmentException {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+        ensureInitializedAndNotClosed();
+
+        manager.seal(handle.getSegmentName());
+    }
+
+    @Override
+    public void unseal(SegmentHandle handle) {
+        throw new UnsupportedOperationException("Unseal is not implemented for BookKeeperStorage");
```

Comment:
What's the plan for these supported operations?

Reply:
I assume you mean *un*suported?

Truncate is optional. RollingStorage takes care of that.
`unseal` is a nice to have, but not required.